File: grafos_UI/grafos_ui/controller.py
from __future__ import annotations

import random
from collections.abc import Sequence

from .algorithms import (
    BreadthFirstStepper,
    DepthFirstStepper,
    DijkstraStepper,
    RandomWalkStepper,
    SearchStepper,
)
from .domain import GenerationMetadata, GraphConfig, GridGraph, SearchSnapshot
from .generator import GridGraphFactory


class SimulationController:
    def __init__(
        self,
        graph_factory: GridGraphFactory | None = None,
        stepper_types: Sequence[type[SearchStepper]] | None = None,
    ) -> None:
        self.graph_factory = graph_factory or GridGraphFactory()
        self.stepper_types = stepper_types or (
            BreadthFirstStepper,
            DepthFirstStepper,
            DijkstraStepper,
            RandomWalkStepper,
        )
        self.config: GraphConfig | None = None
        self.graph: GridGraph | None = None
        self.metadata: GenerationMetadata | None = None
        self.running = False
        self._seed_anchor = random.SystemRandom().randint(0, 10**9)
        self._steppers: dict[str, SearchStepper] = {}

    def generate_graph(self, config: GraphConfig) -> GridGraph:
        self.pause()
        self.config = config
        self.graph, self.metadata = self.graph_factory.build(config)
        self._seed_anchor = config.seed if config.seed is not None else self._seed_anchor + 97
        self._steppers = {}
        for index, stepper_type in enumerate(self.stepper_types):
            seed = self._seed_anchor + index
            self._steppers[stepper_type.algorithm_id] = stepper_type(self.graph, seed=seed)
        return self.graph

    def start(self) -> None:
        if self.graph is not None:
            self.running = True

    def pause(self) -> None:
        self.running = False

    def reset(self) -> None:
        self.pause()
        for stepper in self._steppers.values():
            stepper.reset()

    def step_all(self) -> dict[str, SearchSnapshot]:
        snapshots = {name: stepper.step() for name, stepper in self._steppers.items()}
        if snapshots and all(stepper.is_finished for stepper in self._steppers.values()):
            self.running = False
        return snapshots

    def snapshots(self) -> dict[str, SearchSnapshot]:
        return {name: stepper.snapshot() for name, stepper in self._steppers.items()}

    @property
    def steppers(self) -> dict[str, SearchStepper]:
        return dict(self._steppers)

File: grafos_cpp/src/controller.cpp
#include "grafos_cpp/controller.hpp"

#include <random>

namespace grafos {

SimulationController::SimulationController() : seed_anchor_(random_seed()) {}

const GridGraph& SimulationController::generate_graph(const GraphConfig& config) {
    pause();
    config_ = config;
    auto build = graph_factory_.build(config);
    graph_ = std::move(build.graph);
    metadata_ = std::move(build.metadata);
    seed_anchor_ = config.seed.value_or(seed_anchor_ + 97U);
    rebuild_steppers();
    return *graph_;
}

void SimulationController::start() {
    if (graph_.has_value()) {
        running_ = true;
    }
}

void SimulationController::pause() {
    running_ = false;
}

void SimulationController::reset() {
    pause();
    for (auto& [_, stepper] : steppers_) {
        stepper->reset();
    }
}

std::map<std::string, SearchSnapshot> SimulationController::step_all() {
    std::map<std::string, SearchSnapshot> snapshots;
    for (auto& [name, stepper] : steppers_) {
        snapshots[name] = stepper->step();
    }

    if (!snapshots.empty()) {
        bool all_finished = true;
        for (const auto& [_, stepper] : steppers_) {
            if (!stepper->is_finished()) {
                all_finished = false;
                break;
            }
        }
        if (all_finished) {
            running_ = false;
        }
    }
    return snapshots;
}

std::map<std::string, SearchSnapshot> SimulationController::snapshots() const {
    std::map<std::string, SearchSnapshot> snapshots;
    for (const auto& [name, stepper] : steppers_) {
        snapshots[name] = stepper->snapshot();
    }
    return snapshots;
}

bool SimulationController::running() const {
    return running_;
}

const GraphConfig* SimulationController::config() const {
    return config_.has_value() ? &*config_ : nullptr;
}

const GridGraph* SimulationController::graph() const {
    return graph_.has_value() ? &*graph_ : nullptr;
}

const GenerationMetadata* SimulationController::metadata() const {
    return metadata_.has_value() ? &*metadata_ : nullptr;
}

unsigned int SimulationController::random_seed() {
    std::random_device device;
    return device();
}

void SimulationController::rebuild_steppers() {
    steppers_.clear();
    if (!graph_.has_value()) {
        return;
    }

    steppers_["bfs"] = std::make_unique<BreadthFirstStepper>(*graph_, seed_anchor_ + 0U);
    steppers_["dfs"] = std::make_unique<DepthFirstStepper>(*graph_, seed_anchor_ + 1U);
    steppers_["dijkstra"] = std::make_unique<DijkstraStepper>(*graph_, seed_anchor_ + 2U);
    steppers_["random"] = std::make_unique<RandomWalkStepper>(*graph_, seed_anchor_ + 3U);
}

}  // namespace grafos
